### PR TITLE
[WIP] Rootfinding javier

### PR DESCRIPTION
**Pull request by JavierSijen**
I wanted to redo old PPM, fixed-point calculations, but I couldn't in the new branch. Why not re-implement it and keep it as an additional rootfinder? Also doing some refactoring of the rootfinding part.

### CODE REVIEW

**Comment by JensWehner** [comment]
The old fixed point is simply not a as good as Newton while less stable than the grid so I do not see the point

**Comment by JavierSijen** [comment]
> 
> 
> The old fixed point is simply not a as good as Newton while less stable than the grid so I do not see the point


> 
> 
> The old fixed point is simply not a as good as Newton while less stable than the grid so I do not see the point

You're probably right. I do think the rootfinding stuff could be refactored slightly. This branch contains some nice refactorings imo.

File: src/libxtp/gwbse/gw.cc [hunk 1]
```diff
@@ -210,29 +219,35 @@ Eigen::VectorXd GW::SolveQP(const Eigen::VectorXd& frequencies) const {
...
   for (Index gw_level = 0; gw_level < _qptotal; ++gw_level) {
-    double initial_f = frequencies[gw_level];
+    double frequency = frequencies[gw_level];
     double intercept = intercepts[gw_level];
-    boost::optional<double> newf;
+    QPFunc fqp(gw_level, *_sigma.get(), intercept);
+    boost::optional<double> newf = boost::none;
+    // Call solver of choice
     if (_opt.qp_solver == "fixedpoint") {
-      newf = SolveQP_FixedPoint(intercept, initial_f, gw_level);
```
Comment (JensWehner) [inline src/libxtp/gwbse/gw.cc]: yeah I do not want the fixed point solver anymore.

File: src/libxtp/gwbse/gw.cc [hunk 2]
```diff
@@ -251,111 +266,103 @@ Eigen::VectorXd GW::SolveQP(const Eigen::VectorXd& frequencies) const {
   return frequencies_new;
 }
 
-boost::optional<double> GW::SolveQP_Linearisation(double intercept0,
-                                                  double frequency0,
-                                                  Index gw_level) const {
+boost::optional<double> GW::SolveQP_Bisection(double freq_lb, double freq_ub,
```
Comment (JensWehner) [inline src/libxtp/gwbse/gw.cc]: Bisection is guaranteed to converge so why return an optional?

File: src/libxtp/gwbse/gw.cc [hunk 3]
```diff
@@ -251,111 +266,103 @@ Eigen::VectorXd GW::SolveQP(const Eigen::VectorXd& frequencies) const {
...
+      break;
+    }
+    freq_prev = freq_curr;
   }
   return newf;
 }
 
-boost::optional<double> GW::SolveQP_Grid(double intercept0, double frequency0,
-                                         Index gw_level) const {
+boost::optional<double> GW::SolveQP_Grid(double frequency0,
```
Comment (JensWehner) [inline src/libxtp/gwbse/gw.cc]: nice

File: src/libxtp/gwbse/gwbse.cc [hunk 4]
```diff
@@ -394,13 +394,25 @@ void GWBSE::Initialize(tools::Property& options) {
   }
   XTP_LOG(Log::error, *_pLog) << " eta: " << _gwopt.eta << flush;
 
-  _gwopt.qp_solver = options.ifExistsReturnElseReturnDefault<std::string>(
-      key + ".qp_solver", _gwopt.qp_solver);
+  if (options.exists(key + ".qp_solver")) {
+    std::vector<std::string> solvers = {"fixedpoint", "grid", "newton"};
```
Comment (JensWehner) [inline src/libxtp/gwbse/gwbse.cc]: yeah, please leave fixed point out.

File: src/libxtp/gwbse/gw.cc (hunk 2 above)
Comment (JavierSijen) [reply src/libxtp/gwbse/gw.cc]: In this branch, bisection could fail if the bisection loop counter exceeds `_opt.g_sc_max_iterations`

File: src/libxtp/gwbse/gw.cc (hunk 1 above)
Comment (JavierSijen) [reply src/libxtp/gwbse/gw.cc]: Okay!

**Comment by JavierSijen** [comment]
Thanks for the comments, Jens! I'll keep the branch for personal use, but it doesn't need to be merged into master, as far as I am concerned.

**Comment by JensWehner** [comment]
@JavierSijen is this still relevant or can it be deleted?

**Comment by JavierSijen** [comment]
> 
> 
> @JavierSijen is this still relevant or can it be deleted?

This can be deleted!

**Pull request closed without merging**